### PR TITLE
spi: add a SPI driver for the SAM0 series

### PR DESCRIPTION
**Pull request by nzmichaelh**
This patch adds a master only, polling based SPI driver for the SAM0 series and enables it for the built-in port on the Arduino Zero.

It includes special cases for the common write-only, read-only, and half duplex cases.  With these the driver can keep the bus >= 95 % utilised at SCK=12 MHz and >= 75 % utilised at SCK=24 MHz.

In future patches I plan to add a SPI Device Tree binding and wire it in, and then to add DMA support.

### CODE REVIEW

**Review by dbkinder — APPROVED** [review]
+1 for docs

File: drivers/spi/spi_sam0.c [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+/* Fast path that transmits a buf */
+static void spi_sam0_fast_tx(SercomSpi *regs, const struct spi_buf *tx_buf)
+{
+	const u8_t *p = tx_buf->buf;
+	const u8_t *pend = tx_buf->buf + tx_buf->len;
+	u8_t ch;
+
+	while (p != pend) {
+		ch = *p++;
+		while (!regs->INTFLAG.bit.DRE) {
```
Comment (tbursztyka) [inline drivers/spi/spi_sam0.c]: break a line before the while

File: drivers/spi/spi_sam0.c [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+		}
+		regs->DATA.reg = ch;
```
Comment (tbursztyka) [inline drivers/spi/spi_sam0.c]: and before this line as well. Actually, apply this comment everywhere: it's ugly to get a statement glued to a previous line with }

File: drivers/spi/spi_sam0.c [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+	if (len <= 0) {
+		return;
+	}
+
+	/* Load the first outgoing byte */
+	while (!regs->INTFLAG.bit.DRE) {
+	}
+	regs->DATA.reg = 0;
+
+	while (len--) {
```
Comment (tbursztyka) [inline drivers/spi/spi_sam0.c]: I'd rather do a while (len) {

File: drivers/spi/spi_sam0.c [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+		if (len != 0) {
```
Comment (tbursztyka) [inline drivers/spi/spi_sam0.c]: remove that if (keep the code inside however)

File: drivers/spi/spi_sam0.c [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+			while (!regs->INTFLAG.bit.DRE) {
+			}
+			regs->DATA.reg = 0;
+		}
+
+		while (!regs->INTFLAG.bit.RXC) {
+		}
+
+		*p++ = regs->DATA.reg;
```
Comment (tbursztyka) [inline drivers/spi/spi_sam0.c]: and put a len--; after this line.

it gains one test per loop in the end.

File: drivers/spi/spi_sam0.c [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+	if (len <= 0) {
+		return;
+	}
+
+	/* Load the first outgoing byte */
+	while (!regs->INTFLAG.bit.DRE) {
+	}
+	regs->DATA.reg = *psrc++;
+
+	while (len--) {
```
Comment (tbursztyka) [inline drivers/spi/spi_sam0.c]: do the same as previous function.

File: drivers/spi/spi_sam0.c [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+
+	/* Note that all transmits are complete and the RX buf is empty */
+}
+
+/* Finish any ongoing writes and drop any remaining read data */
+static void spi_sam0_finish(SercomSpi *regs)
+{
+	while (!regs->INTFLAG.bit.TXC) {
+	}
+	while (regs->INTFLAG.bit.RXC) {
```
Comment (tbursztyka) [inline drivers/spi/spi_sam0.c]: break a line before.

File: drivers/spi/spi_sam0.c [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+
+	/* The device will be configured and enabled when transceive
+	 * is called.
+	 */
+
+	return 0;
+}
+
+static const struct spi_driver_api spi_sam0_driver_api = {
+	.transceive = spi_sam0_transceive, .release = spi_sam0_release,
```
Comment (tbursztyka) [inline drivers/spi/spi_sam0.c]: one assignment per line please.

File: drivers/spi/spi_sam0.c [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+{
+	const struct spi_sam0_config *cfg = config->dev->config->config_info;
+	struct spi_sam0_data *data = config->dev->driver_data;
+	SercomSpi *regs = cfg->regs;
+	int err;
+
+	spi_context_lock(&data->ctx, false, NULL);
+
+	err = spi_sam0_configure(config);
+
```
Comment (tbursztyka) [inline drivers/spi/spi_sam0.c]: you can remove that empty line as you are testing the previous assignment  in the if below.

**Review by tbursztyka — CHANGES_REQUESTED** [review]
There is no possibility to make this driver interrupt based?

**Comment by nzmichaelh** [comment]
@tbursztyka re: interrupt based, the SAM0 has no FIFO so you'd spend as much time servicing interrupts as polling.

I'm planning to do a follow up PR that adds a DMA driver and then switches this over to DMA.  That'll solve the CPU usage problem.

File: drivers/spi/spi_sam0.c (hunk 2 above)
Comment (nzmichaelh) [reply drivers/spi/spi_sam0.c]: Done globally.

File: drivers/spi/spi_sam0.c (hunk 4 above)
Comment (nzmichaelh) [reply drivers/spi/spi_sam0.c]: This is a performance optimisation that interleaves the transmit and receive operations.  The logic is:

Start transmit of byte 0

Loop:
  Start transmit of byte n+1
  Receive byte n

I've added a comment to patch.  SG?

File: drivers/spi/spi_sam0.c (hunk 5 above)
Comment (nzmichaelh) [reply drivers/spi/spi_sam0.c]: Done as part of earlier comments.

File: drivers/spi/spi_sam0.c (hunk 6 above)
Comment (nzmichaelh) [reply drivers/spi/spi_sam0.c]: Ack.  Depends on review of the previous function.

File: drivers/spi/spi_sam0.c (hunk 9 above)
Comment (nzmichaelh) [reply drivers/spi/spi_sam0.c]: Done.

File: drivers/spi/spi_sam0.c (hunk 8 above)
Comment (nzmichaelh) [reply drivers/spi/spi_sam0.c]: Done.

**Comment by nzmichaelh** [comment]
@tbursztyka thanks for the review!  PTAL.

**Comment by pfalcon** [comment]
@sknsean: Please help to review this (and other SAM patches). Thanks.

**Review by tbursztyka — APPROVED** [review]

**Comment by sknsean** [comment]
Please consider what I'm doing in https://github.com/zephyrproject-rtos/zephyr/pull/5623
There are other SAMD's that just the SAMD21 included here :-)
Will do a further review on this PR from a SAMD20's perspective.

**Comment by nzmichaelh** [comment]
Hi @sknsean the SAMD20 is part of the same SAM0 series so this driver should work as-is.  Let me know how it goes.

**Comment by skny-prevas** [comment]
Hi @nzmichaelh not exactly see here [1] what changes I needed in RIOT OS.
[1] https://github.com/sknsean/RIOT/commit/9fab75e7977994dfdb4e8a9f06e0158cbdd9cee1

**Comment by sknsean** [comment]
Hi @nzmichaelh not exactly see here [1] what changes I needed in RIOT OS.
[1] sknsean/RIOT@9fab75e

**Comment by nzmichaelh** [comment]
@sknsean oh, that's crazy!  Why did they do that?  I can understand the SAMLxx and SAMDxx having differences, but why the SAMD21 and SAMD20...

Done using feature test macros.

**Comment by sknsean** [comment]
Really crazy :-) I guess that's just Atmel's way of complicating things.

**Pull request merged**